### PR TITLE
Allow connecting nested collections to multiple=true inputs in workflow editor

### PR DESCRIPTION
**Pull request by mvdbeek**
screen shot 2019-02-28 at 22 05 30

### CODE REVIEW

**Comment by innovate-invent** [comment]
Should I go ahead and patch my Galaxy instance with this or is it still WIP?

**Comment by mvdbeek** [comment]
Sure, give it a try. I haven't tested this yet beyond being able to create the workflow in the editor.

**Comment by innovate-invent** [comment]
Just rigging up my workflow now, it seems to delete the list:list connection every time I change a tool parameter

**Comment by innovate-invent** [comment]
Shoot, this change allows the connection but Galaxy seems to flatten the list before passing it to the tool.

**Comment by mvdbeek** [comment]
We fixed that recently, and I think we also backported that to 18.09

**Comment by mvdbeek** [comment]
Nope, just 19.01.

**Comment by innovate-invent** [comment]
Shoot, ok, looks like I am upgrading...

**Comment by mvdbeek** [comment]
> it seems to delete the list:list connection every time I change a tool parameter

It should only delete the connection if the list type changes, so for instance if you produce `list` or `list:list:list` instead of `list:list`. Simple parameter changes don't change the downstream connection for me. I've played around with this some more and I think it works fine, so taking it out of WIP.

**Comment by mvdbeek** [comment]
damn, no travis here ...

**Comment by innovate-invent** [comment]
Sorry, to be clear, the tool receiving the list:list breaks its input connection when you change its parameters. This was tested on 18.09 though, I am trying to set up a 19.01 instance.

**Comment by mvdbeek** [comment]
What tool is that, and what parameter change causes this?

**Comment by innovate-invent** [comment]
iuc progressiveMauve, any of the toggles.

**Comment by innovate-invent** [comment]
Just loaded a 19.01 instance with the patch and I get this:
image

It isnt a draw artifact, moving the outputting tool will update the ghost line to follow it.
Creating the connection causes my browser to freeze up for a second when I release the mouse button. Saving and reloading with this artifact also locks up the browser, its weird though, the webpage is unresponsive except for when hovering over the used space info in the top left. It still pops up the tool tip describing it.

Changing the parameters of the tool doesn't break the connection though.

**Comment by mvdbeek** [comment]
Yeah, probably a javascript error, can you open the browser console and see if there are any errors ?

**Comment by innovate-invent** [comment]
```
TypeError: this.terminalMapping is undefined[Learn More] libs.chunk.js:182:7 
	setMapOver workflow-terminals.js:171 connect workflow-terminals.js:403 connect workflow-connector.js:23 from_simple workflow-manager.js:266 each jQuery from_simple workflow-manager.js:263 each jQuery from_simple workflow-manager.js:258 each jQuery from_simple workflow-manager.js:256 success workflow-view.js:333 fire4jQuery
```

**Comment by mvdbeek** [comment]
I do see the disconnect as well, and it shouldn't happen AFAICT, but the error you posted I can't reproduce, any chance something went wrong building the client, or the browser still using an old version of this ? The changes in workflow-view-terminals.js should prevent this error from occurring.

**Comment by innovate-invent** [comment]
Here is the console output while adding and attaching the list:list between the tools.
```
Polyfills are running polyfills.js:28
JQMIGRATE: Migrate is installed with logging active, version 1.4.1 jquery-migrate.js:26
setGalaxyInstance /workflow/editor singleton.js:10
You are running Vue in development mode.
Make sure to turn on production mode when deploying for production.
See more tips at https://vuejs.org/guide/deployment.html vue.runtime.esm.js:8021
setGalaxyInstance /workflow/editor singleton.js:10
unreachable code after return statement[Learn More] libs.chunk.js:142:4
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/ libs.chunk.js:9064:4
Source map error: request failed with status 404
Resource URL: http://vm:9080/static/style/base.css?v=1551459098
Source Map URL: base.css.map[Learn More]
Warning: No collection type or collection type source defined. workflow-terminals.js:555
JQMIGRATE: Can't change the 'type' of an input or button in IE 6/7/8 jquery-migrate.js:48
console.trace() jquery-migrate.js:50

    migrateWarn4jQuery
    render ui-misc.js:94 triggerEvents5Backbone
    render form-input.js:129 initialize form-input.js:49 View2Backbone
    _addRow form-section.js:171 add form-section.js:44 render form-section.js:24 forEach Underscore render form-section.js:23 initialize form-section.js:16 View2Backbone
    _renderForm form-view.js:132 render form-view.js:102 _render tool-form-base.js:89 _update tool-form-base.js:48 then8jQuery
    _update tool-form-base.js:47 execute deferred.js:44 fire3jQuery
    execute deferred.js:41 _update tool-form-base.js:45 initialize tool-form-base.js:25 View3Backbone
    initialize workflow-forms.js:44 View2Backbone
    showForm workflow-view.js:760 node_changed workflow-manager.js:327 init_field_data workflow-node.js:230 success workflow-view.js:648 request utils.js:192 fire4jQuery

Warning: No collection type or collection type source defined. workflow-terminals.js:568
TypeError: this.terminalMapping is undefined[Learn More] libs.chunk.js:182:7 
```

**Comment by mvdbeek** [comment]
I fixed the issue with outputs getting disconnected, but I don't see the error you're seeing.
connect_and_change_values

**Comment by mvdbeek** [comment]
Also I'm not sure all of this is going to work on 19.01, there were some additional changes that might be missing if you just add the changes here.

**Comment by innovate-invent** [comment]
So I will have to move to dev?

**Comment by mvdbeek** [comment]
Maybe try this one again with the new commit added. Also make sure you really got the change in line 80 of `client/galaxy/scripts/mvc/workflow/workflow-view-terminals.js`.

**Comment by innovate-invent** [comment]
I added your most recent commit, rebuilt, and everything worked perfectly. Thanks so much for fixing this.

File: client/galaxy/scripts/mvc/workflow/workflow-terminals.js [hunk 1]
```diff
@@ -157,14 +157,21 @@ var Terminal = Backbone.Model.extend({
...
+        let output_val = val;
         if (this.multiple) {
-            return; // Cannot set this to be multirun...
+            // emulate list input
+            let description = new CollectionTypeDescription('list');
+            if (val.collectionType === description.collectionType) {
+                // No mapping over necessary
+                return;
+            }
+            output_val = val.effectiveMapOver ? val.effectiveMapOver(description): val;
```
Comment (jmchilton) [inline client/galaxy/scripts/mvc/workflow/workflow-terminals.js:168]: Nifty.

**Pull request merged**